[TASK] Add sql scripts

MatthewPhinney opened the pull request.

MatthewPhinney commented:
I need to refactor this to use the approach in https://github.com/interledger/five-bells-ledger/pull/256. Also, I want to review the database design. It doesn't conform exactly to Abraham's specifications.

MatthewPhinney commented on src/lib/db.js:
```diff
@@ -0,0 +1,59 @@
+'use strict'
```
I think this file is identical to the one in the ledger. We may want to consolidate in `five-bells-shared` at some point.

MatthewPhinney commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+-- Initial Lookup values
+--
+
+INSERT INTO "N_LU_CASE_STATUS" ("CASE_STATUS_ID", "CASE_STATUS_CODE", "CASE_STATUS_DESC") VALUES (1, 'proposed', 'This case is proposed');
```
This has been architected with a lookup table in which integer keys map to descriptions. One of my concerns is that this leads to brittle code. If someone changes the db script, but forgets to update `db/case.js` or `db/notification.js`, the numeric keys will no longer correspond to the correct case/notification states. I'm still thinking about how to make this more robust.

scratch28 replied:
if the code was dependent on the LU table for logic and values then it would not be so “brittle”
this also is helpful if future environments (qa, perf, prod) do not assume the surrogate key values and use the real values as persisted in db

nararama replied:
To prevent someone from changing the script, We have to have the  primary key not null defined for the  look up table . For controlling the .js code you have to write a select statement in the .js  code and check if the record exist in the table. If it exist raise an exception and move on.

MatthewPhinney replied:
I updated the methods that convert to/from persistent in the js code.

MatthewPhinney commented:
I added drop scripts, but they are not yet used by the code. I think codecov is complaining because I added sql scripts that do not have test coverage.

clark800 commented on src/models/db/notification.js:
```diff
@@ -1,26 +1,89 @@
+  return !options ? knex : (!options.transaction ? knex : options.transaction)
+}
+
+function saveNotification (notificationObject, options) {
```
I think we always know whether we are updating or inserting, so I think we can split this into two methods.

clark800 commented on src/models/cases.js:
```diff
@@ -1,35 +1,82 @@
+}
+
+CasesFactory.constitute = [NotificationWorker, CaseExpiryMonitor]
+function CasesFactory (notificationWorker, caseExpiryMonitor) {
```
I think it would be easy to flatten this into functions at this point right?

MatthewPhinney replied:
Last time I tried, I ran into trouble with some of the required modules. I'll take another look.

clark800 commented on src/models/cases.js:
```diff
@@ -41,55 +88,55 @@ function CasesFactory (Case, log, notificationWorker, caseExpiryMonitor) {
+        if (caseInstance.case_status_id !== 'executed') {
+          caseInstance.case_status_id = 'executed'
           caseInstance.exec_cond_fulfillment = fulfillment
-          yield db.updateCase(caseInstance, {transaction})
```
Maybe we should keep `updateCase` available since it would probably be faster

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,134 @@
-      }
-      return data
-    }
+function knexOrTransaction (knex, options) {
```
I renamed this to `getTransaction` in the ledger, maybe not the perfect name, but I think it's better than `knexOrTransaction`

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,134 @@
+      } else if (results.length === 0) {
+        return null
+      } else {
+        assert(false)
```
Helpful to include an error message in the assert

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,134 @@
+    })
+}
+
+function getCaseByUuid (caseUuid, options) {
```
Can reduce duplication by using a `getCaseWhere` function

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,134 @@
-      data.notaries = data.notaries[0]
-      return data
-    }
+function save (caseObject, options) {
```
Same here, probably better to split into insert and update methods if possible

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,134 @@
 
-  PersistentKnexModelMixin(Case, knex)
+function convertFromPersistent (data) {
+  data = _.cloneDeep(data)
```
I started trying to avoid reassignment here and making it more functional, not a big issue though

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,134 @@
+  data = _.cloneDeep(data)
+  data = _.mapKeys(data, (value, key) => key.toLowerCase())
+  data.case_status_id = CASE_STATUS_IDS[data.case_status_id]
+  if (!Array.isArray(data.notaries)) {
```
Can we just require it to always be an array?

MatthewPhinney replied:
I'm not sure why the database stores the first notary in the array, while the JSON allows multiple notaries in the array. Looks like Nao added the code originally, and he's not around to ask. Before I change this, I'd like to check with @stefan to determine the use case for multiple notaries per case.

MatthewPhinney commented:
Requires 5B shared version >16.1.0

MatthewPhinney commented on src/models/cases.js:
```diff
@@ -1,35 +1,83 @@
+}
+
+CasesFactory.constitute = [NotificationWorker, CaseExpiryMonitor]
+function CasesFactory (notificationWorker, caseExpiryMonitor) {
```
@clark800 it is possible to get rid of the class and make this purely functional, but I'm getting some weird `constitute` dependency issues. I'd prefer to address this in a separate PR, since it's not central to the task of refactoring the db.

MatthewPhinney commented on src/models/db/case.js:
```diff
@@ -1,72 +1,115 @@
+    EXECUTION_CONDITION_DIGEST: data.execution_condition,
+    EXEC_COND_FULFILLMENT: data.exec_cond_fulfillment,
+    EXPIRES_DTTM: data.expires_at,
+    NOTARIES: data.notaries[0],
```
We pick the first notary in the array to persist in the DB, and then convert back to an array when we read from the DB. I'm inclined to leave as is for now. This code has been in here for a while, and it's not related to this PR, so I'd prefer to address any cleanup in a separate task.

clark800 commented on src/lib/notificationWorker.js:
```diff
@@ -121,7 +119,11 @@ class NotificationWorker {
     } else {
-      yield notification.destroy()
+      const sucessfulNotification = _.merge({}, notification, {
+        is_active: +false, // convert to int for database
```
We should move all of these type conversions to the database layer

MatthewPhinney replied:
Good point, let me update

clark800 commented on src/models/cases.js:
```diff
@@ -1,35 +1,83 @@
+  if (data.id) {
+    caseObject.id = uri.parse(data.id, 'case').id.toLowerCase()
+  }
+  data = caseObject
```
This doesn't do anything

clark800 commented on src/models/cases.js:
```diff
@@ -41,14 +89,19 @@ function CasesFactory (Case, log, notificationWorker, caseExpiryMonitor) {
       }
 
-      const existed = yield db.upsertCase(Case, caseInstance)
+      const existingCase = yield db.getCaseByUuid(caseInstance.id)
```
This block could be moved into an `upsert` method in the database layer

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,115 @@
+  return knex.transaction(co.wrap(callback))
+}
+
+function * insertCases (caseObject, options) {
```
Should pluralize to `caseObjects` or just `cases`

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+    "DB_UPDATED_USER"            CHARACTER VARYING(40) DEFAULT  CURRENT_USER  NOT NULL 
+);
+
+CREATE INDEX "XPKN_LU_CASE_STATE" ON "N_LU_CASE_STATUS"
```
Based on the email I sent out from the discussion with Abraham, the N is supposed to be separated by an underscore and before the other prefix.

MatthewPhinney replied:
It looks like this is also captured on Confluence: https://ripplelabs.atlassian.net/wiki/display/DT/Database+design+patterns.

clark800 commented on test/helpers/db.js:
```diff
@@ -0,0 +1,39 @@
+      yield knex.raw('TRUNCATE TABLE \"' + t + '\" CASCADE;').then()
+      /* eslint-enable no-useless-escape */
+    } else {
+      yield knex(t).truncate().then()
```
Would be good to move `truncateTables` to `src/lib/db` too to keep direct access to `knex` out of the `test` directory

scratch28 commented:
LGTM

clark800 commented on src/lib/db.js:
```diff
@@ -0,0 +1,94 @@
+function * truncateTables () {
+  for (let t of tables) {
+    if (knex.client.config.client === 'pg') {
+      /* eslint-disable no-useless-escape */
```
Why do you need a useless escape?

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,163 @@
-  Case.validateExternal = validator.create('Case')
-  Case.tableName = 'cases'
+function convertToPersistent (data, options) {
+  const caseStatusRows = getTransaction(options)(STATUS_TABLE_NAME)
```
Can we just query this table once per application instantiation?

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,163 @@
+    return
+  }
+  return Promise.all(_.map(caseObjects, (caseObject) => {
+    return convertToPersistent(caseObject, options)
```
This doesn't have to be inside the promise chain right? Should be simpler to convert first and then use promises just for the inserts.

clark800 replied:
Oh, this is assuming we can get rid of the query in convertToPeristent

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,163 @@
+  }
+}
+
+function getByKey (key, value, options) {
```
It might be worth using this helper module to separate out the lower-level database code: https://github.com/interledger/five-bells-ledger/blob/master/src/models/db/utils.js

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+CREATE TABLE "N_LU_CASE_STATUS"
+(
+    "CASE_STATUS_ID"             INTEGER NOT NULL ,
+    "CASE_STATUS_CODE"           CHARACTER VARYING(20) NOT NULL ,
```
Here is an example of the prefixes I was mentioning earlier. I think we were going to go with calling these just "CODE" and "DESCRIPTION"

MatthewPhinney replied:
Is this documented on the confluence page? I didn't see it.

clark800 replied:
Oh, looks like that one was missed in the email, but it's something I talked to Abraham about. I wanted to avoid having an arbitrary boundary where some column names were prefixed with the table name and others aren't based on a vague notion of how common the column name is. We decided that only putting the table name on the ID columns was a good option.

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+    "NOTIFICATION_ID"            INTEGER NOT NULL DEFAULT nextval('"N_NOTIFICATION_ID_SEQ"'::regclass) ,
+    "NOTIFICATION_STATUS_ID"     INTEGER NOT NULL REFERENCES "N_LU_NOTIFICATION_STATUS"("NOTIFICATION_STATUS_ID"),
+    "CASE_ID"                    INTEGER NOT NULL REFERENCES "N_CASES"("CASE_ID") ,
+    "NOTIFICATION_TARGET"        CHARACTER VARYING(4000) NULL ,
```
This can probably be "TARGET"

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+ALTER TABLE ONLY "N_CASES"
+    ADD CONSTRAINT "N_PK_CASES" PRIMARY KEY ("CASE_ID");
+
+CREATE UNIQUE INDEX "N_XAK1CASES_GUID" ON "N_CASES"
```
I think one of the rules in the email was that we can drop the enumerations, so for example this would be `N_XAK_CASES_GUID`

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+CREATE INDEX "N_XPK_NOTIFICATIONS" ON "N_NOTIFICATIONS"
+("NOTIFICATION_ID"   DESC);
+
+CREATE INDEX "N_XIF1_NOTIFICATIONS" ON "N_NOTIFICATIONS"
```
In this case I think we were gonna do something like "N_XIF_NOTIFICATIONS_CASES" and "N_XIF_NOTIFICATIONS_STATUS"

clark800 commented on src/sql/sql-server/create.sql:
```diff
@@ -0,0 +1,183 @@
+)
+go
+
+CREATE UNIQUE CLUSTERED INDEX XPKN_LU_CASE_STATE ON N_LU_CASE_STATUS
```
`N_XPK_LU_CASE_STATE`

MatthewPhinney replied:
I haven't change the sql server scripts to conform to the new standards, since we don't yet support sql server. I left a comment explaining this at the top of the file. I added them for completeness, but I can remove them if it's confusing.

clark800 replied:
Oh, I didn't know that you added the mssql files, I think we should omit anything related to mssql

clark800 commented on src/sql/sql-server/create.sql:
```diff
@@ -0,0 +1,183 @@
+CREATE TABLE N_CASES
+( 
+  CASE_ID              bigint  NOT NULL 
+       constraint DF_n_case_id DEFAULT next value for seq_n_cases_pk,
```
All names should be capitalized

clark800 commented on src/sql/sql-server/create.sql:
```diff
@@ -0,0 +1,183 @@
+)
+go
+
+CREATE UNIQUE CLUSTERED INDEX XPKN_CASES ON N_CASES
```
`N_XPK_CASES`

clark800 commented on src/sql/sql-server/create.sql:
```diff
@@ -0,0 +1,183 @@
+)
+go
+
+CREATE UNIQUE CLUSTERED INDEX XPKLU_NOTIFICATION_STATUS ON N_LU_NOTIFICATION_STATUS
```
`N_XPK_LU_NOTIFICATION_STATUS`

MatthewPhinney commented on src/sql/strong-oracle/drop.sql:
```diff
@@ -0,0 +1,154 @@
+DROP TRIGGER tU_cases;
+/
+
+DROP TRIGGER tI_cases;
```
@scratch28 I noticed there are some triggers in the drop script (like this one) that are not in the create script. Is there a reason for this?

scratch28 replied:
these should be removed because you altered the name of the create triggers to something else and in the drop you followed that name
these are the old names that I originally proposed

scratch28 commented on src/sql/strong-oracle/drop.sql:
```diff
@@ -0,0 +1,154 @@
+DROP INDEX XPKLU_NOTIFICATION_STATUS;
+/
+
+DROP TRIGGER tD_lu_notification_state;
```
missing create for this trigger tD_lu_notification_state, tI_lu_notification_state , and tU_lu_notification_state

MatthewPhinney commented:
@clark800 I updated the way we read from the lookup tables. It's done on application startup. Then we use the values stored in memory for the convert to/from persistent methods. I also updated the naming conventions of the SQL triggers/columns/indexes/sequences. I think it matches the specifications now. 

As for using [this file](https://github.com/interledger/five-bells-ledger/blob/4af71a3f490ad2f1766f0d5c540c9ad0d514bf8b/src/models/db/utils.js), I'm inclined to hold off, since there are really only two db tables, and this PR is pretty big already. I think we can re-asses if we want to refactor to use that file once this is merged.

clark800 commented on src/models/db/case.js:
```diff
@@ -1,72 +1,141 @@
+  return getByKey('UUID', caseUuid, options)
+}
+
+function convertFromPersistent (data, options) {
```
`options` is unused

MatthewPhinney replied:
Oops I forgot to delete this.

clark800 commented on src/models/db/caseStatus.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+function getStatusDescription (statusCode) {
+  return CASE_STATUS_CODES[statusCode]
```
Looks like these are messed up

MatthewPhinney replied:
The logic is correct, I think the variable names are what is confusing. Let me think of better ones.

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+    "DB_UPDATED_USER"   CHARACTER VARYING(40) DEFAULT  CURRENT_USER  NOT NULL 
+);
+
+CREATE INDEX "N_XPK_LU_CASE_STATE" ON "N_LU_CASE_STATUS"
```
`STATE` should be `STATUS`?

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+ALTER TABLE ONLY "N_CASES"
+    ADD CONSTRAINT "N_PK_CASES" PRIMARY KEY ("CASE_ID");
+
+CREATE UNIQUE INDEX "N_XAKCASES_GUID" ON "N_CASES"
```
`N_XAK_CASES_GUID`

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+    "DB_UPDATED_USER"            CHARACTER VARYING(40) DEFAULT  USER  NOT NULL 
+);
+
+CREATE UNIQUE INDEX "N_XPKLU_NOTIFICATION_STATUS" ON "N_LU_NOTIFICATION_STATUS"
```
`N_XPK_LU_NOTIFICATION_STATUS`

clark800 commented on src/sql/pg/create.sql:
```diff
@@ -0,0 +1,147 @@
+CREATE TABLE "N_LU_CASE_STATUS"
+(
+    "CASE_STATUS_ID"    INTEGER NOT NULL ,
+    "STATUS_CODE"       CHARACTER VARYING(20) NOT NULL ,
```
I think the convention would make these `CODE` and `DESCRIPTION`

clark800 commented on src/sql/strong-oracle/create.sql:
```diff
@@ -0,0 +1,287 @@
+CREATE SEQUENCE SEQ_N_CASES_PK
```
`N_` prefix should be leftmost.

MatthewPhinney commented:
@clark800 updated

clark800 commented:
LGTM
